modules/cms-kit/src/Volo.CmsKit.Public.Application/Volo/CmsKit/Public/Menus/MenuPublicAppService.cs

```diff
@@ -1,30 +1,50 @@
+using System.Collections.Generic;
+using System.Linq;
 using System.Threading.Tasks;
+using Volo.Abp.Caching;
 using Volo.Abp.GlobalFeatures;
 using Volo.CmsKit.GlobalFeatures;
 using Volo.CmsKit.Menus;
 
 namespace Volo.CmsKit.Public.Menus
-{    
+{
     [RequiresGlobalFeature(typeof(MenuFeature))]
     public class MenuPublicAppService : CmsKitPublicAppServiceBase, IMenuPublicAppService
     {
         protected IMenuRepository MenuRepository { get; }
 
-        public MenuPublicAppService(IMenuRepository menuRepository)
+        protected IDistributedCache<MenuWithDetailsDto, MainMenuCacheKey> DistributedCache { get; }
+
+        public MenuPublicAppService(
+            IMenuRepository menuRepository,
+            IDistributedCache<MenuWithDetailsDto, MainMenuCacheKey> distributedCache)
         {
             MenuRepository = menuRepository;
+            DistributedCache = distributedCache;
         }
-        
+
         public async Task<MenuWithDetailsDto> GetMainMenuAsync()
         {
-            var menu = await MenuRepository.FindMainMenuAsync(includeDetails: true);
+            var cachedMenu = await DistributedCache.GetOrAddAsync(
+                new MainMenuCacheKey(),
+                async () =>
+                {
+                    var menu = await MenuRepository.FindMainMenuAsync(includeDetails: true);
+                    
+                    if (menu == null)
```

Comment:
If there is no menu in the database, it will always call this factory method again and again, because `DistributedCache.GetOrAddAsync` [calls the factory](https://github.com/abpframework/abp/blob/cfc648bc656da21b73d99578d71d0ac71b56b9c4/framework/src/Volo.Abp.Caching/Volo/Abp/Caching/DistributedCache.cs#L503-L509) if the item is null in the cache.
However, if it doesn't exists in the database, no point to call it again and again (it should cache that info too)